lib/srv/regular/sshserver.go

```diff
@@ -1096,40 +1109,253 @@ func (s *Server) getServerResource() (types.Resource, error) {
 
 // getDirectTCPIPForwarder sets up a connection-level subprocess that handles forwarding connections. Subsequent
 // calls from the same connection context reuse the same forwarder.
-func (s *Server) getDirectTCPIPForwardDialer(ctx *srv.ServerContext) (sshutils.TCPIPForwardDialer, error) {
-	if d, ok := ctx.Parent().GetDirectTCPIPForwardDialer(); ok {
+func (s *Server) getDirectTCPIPForwardDialer(scx *srv.ServerContext) (sshutils.TCPIPForwardDialer, error) {
+	if d, ok := scx.Parent().GetDirectTCPIPForwardDialer(); ok {
 		return d, nil
 	}
 
-	dialerConn, listenerConn, err := uds.NewSocketpair(uds.SocketTypeDatagram)
+	proc, err := s.startForwardingSubprocess(scx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// set up a dial function that sends the address + fd as a unix datagram message. the forwarder subprocess
+	// interprets all such messages in this way, and will dial the specified address and proxy all traffic
+	// to the desired endpoint.
+	dialer := func(addr string) (net.Conn, error) {
+		local, remote, err := uds.NewSocketpair(uds.SocketTypeStream)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer remote.Close()
+
+		remoteFD, err := remote.File()
+		if err != nil {
+			local.Close()
+			return nil, trace.Wrap(err)
+		}
+		defer remoteFD.Close()
+
+		_, _, err = proc.Conn.WriteWithFDs([]byte(addr), []*os.File{remoteFD})
+		if err != nil {
+			local.Close()
+			return nil, trace.Wrap(err)
+		}
+
+		return local, nil
+	}
+
+	// try to register with the parent context.
+	if other, ok := scx.Parent().TrySetDirectTCPIPForwardDialer(dialer); !ok {
+		// another forwarder was concurrently created. this isn't actually a problem, multiple forwarders
+		// being registered is harmless, but it does result in slightly higher resource utilization, so its
+		// preferable to use the existing forwarder and close ours in the background.
+		go proc.Close()
+		return other, nil
+	}
+
+	// successfully registered this dialer, add closer to context.
+	scx.Parent().AddCloser(proc)
+
+	return dialer, nil
+}
+
+// listenTCPIP creates a new listener in the forwarding process.
+func (s *Server) listenTCPIP(scx *srv.ServerContext, addr string) (net.Listener, error) {
```

Comment:
If we returned a `*net.TCPListener` we'd both check that the type of the listener socket is correct and we'd have an ~infallible check for the listening port since we'd know to expect a `*net.TCPAddr` as the return type of `Addr()`, and that has the port as a number.